File: python/microsoft-agents-protocol/microsoft/agents/protocol/client/simplified_client.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

"""
Simplified high-level client API for Agent Protocol.

Provides convenience methods for common operations matching the .NET implementation.
"""

import json
from typing import Optional, Callable, AsyncIterator, Dict, Any
from .agent_protocol_client import AgentProtocolClient
from .client_options import AgentProtocolClientOptions
from .chat_options import ChatOptions
from .conversation import IConversation, Conversation
from .stream_event import StreamEvent


class SimplifiedClient(AgentProtocolClient):
    """
    Simplified Agent Protocol client with high-level convenience methods.

    Extends AgentProtocolClient with methods like:
    - complete_chat: Simple text-in, text-out
    - complete_chat_structured: Full message objects
    - stream_chat: Streaming with text chunk callbacks
    - create_conversation: Stateful multi-turn conversations
    """

    async def complete_chat(
        self, message: str, options: Optional[ChatOptions] = None
    ) -> str:
        """
        Sends a message and returns the complete response as text (simple API).

        Args:
            message: The message to send
            options: Optional chat options including tools and metadata

        Returns:
            The agent's text response
        """
        # Build request
        request = {
            "input": [
                {"role": "user", "contents": [{"kind": "text", "text": message}]}
            ]
        }

        # Add optional fields
        if options:
            if options.agent_id:
                request["agent_id"] = options.agent_id
            if options.metadata:
                request["metadata"] = options.metadata

        # If tools are provided, handle tool execution automatically
        if options and options.tools:
            return await self._complete_chat_with_tools(request, options)

        # Execute run and wait for completion
        response = await self.runs.create_and_wait(request)

        # Extract text from agent response
        return self._extract_text_from_response(response)

    async def complete_chat_structured(
        self, message: Dict[str, Any], options: Optional[ChatOptions] = None
    ) -> Dict[str, Any]:
        """
        Sends a structured message and returns the complete response message.

        Args:
            message: The message to send (as dict representation of ChatMessage)
            options: Optional chat options

        Returns:
            The agent's response message (as dict)
        """
        # Build request
        request = {"input": [message]}

        # Add optional fields
        if options:
            if options.agent_id:
                request["agent_id"] = options.agent_id
            if options.metadata:
                request["metadata"] = options.metadata

        # Execute run and wait for completion
        response = await self.runs.create_and_wait(request)

        # Extract agent response
        output = response.get("output", [])
        if not output:
            return {"role": "agent", "contents": []}

        # Find first agent message
        for msg in output:
            if msg.get("role") == "agent":
                return msg

        return {"role": "agent", "contents": []}

    async def stream_chat(
        self, message: str, on_text_chunk: Callable[[str], None]
    ) -> None:
        """
        Streams a message response with text chunks delivered via callback.

        Args:
            message: The message to send
            on_text_chunk: Callback fired for each text chunk
        """
        # Build request
        request = {
            "input": [
                {"role": "user", "contents": [{"kind": "text", "text": message}]}
            ]
        }

        accumulated_text = ""

        # Stream events
        async for evt in self._stream_run(request):
            # Handle message delta events for text streaming
            if evt["event_type"] in ("message.delta", "message.updated"):
                message_data = evt.get("data", {})
                contents = message_data.get("contents", [])

                # Find text content
                for content in contents:
                    if content.get("kind") == "text":
                        full_text = content.get("text", "")
                        # Calculate new text since last update
                        if full_text.startswith(accumulated_text):
                            new_text = full_text[len(accumulated_text) :]
                            if new_text:
                                on_text_chunk(new_text)
                                accumulated_text = full_text
                        else:
                            # Full replacement
                            on_text_chunk(full_text)
                            accumulated_text = full_text

    def create_conversation(self) -> IConversation:
        """
        Creates a new conversation for maintaining state across multiple messages.

        Returns:
            A conversation instance
        """
        return Conversation(
            self,
            None,
            enable_logging=self._options.enable_logging,
            log_directory=self._options.log_directory
        )

    def resume_conversation(self, thread_id: str) -> IConversation:
        """
        Resumes an existing conversation using a thread ID.

        Args:
            thread_id: The thread ID to resume

        Returns:
            A conversation instance
        """
        return Conversation(
            self,
            thread_id,
            enable_logging=self._options.enable_logging,
            log_directory=self._options.log_directory
        )

    async def _stream_run(self, request: Dict[str, Any]) -> AsyncIterator[Dict[str, Any]]:
        """
        Internal method to stream a run and parse SSE events.

        Args:
            request: Run request

        Yields:
            Parsed SSE events as dicts with 'event_type' and 'data'
        """
        current_event = None
        current_data = None

        async for line in self.runs.create_and_stream(request):
            line = line.strip()

            if not line:
                # Empty line signals end of event
                if current_event and current_data is not None:
                    yield {"event_type": current_event, "data": current_data}
                    current_event = None
                    current_data = None
                continue

            if line.startswith("event:"):
                current_event = line[6:].strip()
            elif line.startswith("data:"):
                data_str = line[5:].strip()
                if data_str:
                    try:
                        current_data = json.loads(data_str)
                    except json.JSONDecodeError:
                        current_data = {"raw": data_str}

        # Yield last event if exists
        if current_event and current_data is not None:
            yield {"event_type": current_event, "data": current_data}

    async def _complete_chat_with_tools(
        self, request: Dict[str, Any], options: ChatOptions
    ) -> str:
        """
        Handles tool execution automatically during streaming.

        Args:
            request: Run request
            options: Chat options with tools

        Returns:
            Final text response
        """
        result_text = ""

        async for evt in self._stream_run(request):
            if evt["event_type"] in ("message.delta", "message.updated"):
                message_data = evt.get("data", {})
                contents = message_data.get("contents", [])

                # Extract text content
                for content in contents:
                    if content.get("kind") == "text":
                        result_text = content.get("text", "")

            elif evt["event_type"] == "run.requires_action":
                # Extract tool calls and execute them
                # TODO: Implement tool execution loop
                # This requires protocol-level support for submitting tool outputs
                pass

        return result_text

    def _extract_text_from_response(self, response: Dict[str, Any]) -> str:
        """
        Extracts text from a run response.

        Args:
            response: Run response

        Returns:
            Extracted text or empty string
        """
        output = response.get("output", [])
        if not output:
            return ""

        # Find first agent message
        for msg in output:
            if msg.get("role") == "agent":
                contents = msg.get("contents", [])
                for content in contents:
                    if content.get("kind") == "text":
                        return content.get("text", "")

        return ""


def create_simplified_client(
    base_url: str, api_key: Optional[str] = None
) -> SimplifiedClient:
    """
    Creates a simplified Agent Protocol client with convenience methods.

    Args:
        base_url: Base URL for the Agent Protocol API
        api_key: Optional API key for authentication

    Returns:
        Configured simplified client
    """
    options = AgentProtocolClientOptions(base_url=base_url, api_key=api_key)
    return SimplifiedClient(options)

File: typescript/packages/agents-protocol-client/src/simplified-client.ts
/**
 * High-level simplified client API for Agent Protocol
 * Provides easy-to-use methods for chat completion and streaming
 */

import type { ChatMessage, ChatRole } from '@microsoft/agents-protocol-abstractions';
import { AgentProtocolClient } from './client';
import type { AgentProtocolClientConfig } from './types';
import type { ChatOptions } from './chat-options';
import type { StreamEvent } from './stream-event';
import { createStreamEvent } from './stream-event';
import { Conversation, type IConversation } from './conversation';

/**
 * Request for creating a run
 */
export interface RunRequest {
  agentId?: string;
  threadId?: string;
  journalId?: string;
  input: ChatMessage[];
  metadata?: Record<string, unknown>;
  webhook?: string;
}

/**
 * Response from a run operation
 */
export interface RunResponse {
  runId: string;
  threadId: string;
  status: string;
  output?: ChatMessage[];
  error?: {
    code: string;
    message: string;
    details?: Record<string, unknown>;
  };
}

/**
 * High-level client for Agent Protocol with simplified API
 */
export class SimplifiedClient {
  private readonly lowLevelClient: AgentProtocolClient;
  private readonly config: AgentProtocolClientConfig;

  /**
   * Creates a new simplified client
   * @param config Client configuration (baseUrl, authToken, etc.)
   */
  constructor(config: AgentProtocolClientConfig) {
    this.config = config;
    this.lowLevelClient = new AgentProtocolClient(config);
  }

  /**
   * Sends a message and returns the complete response as text (simple API)
   * @param message The message to send
   * @param options Optional chat options (agent ID, tools, metadata)
   * @param signal Optional abort signal for cancellation
   * @returns The agent's text response
   */
  async completeChat(
    message: string,
    options?: ChatOptions,
    signal?: AbortSignal
  ): Promise<string> {
    const request: RunRequest = {
      agentId: options?.agentId,
      input: [
        {
          role: 'user' as ChatRole,
          messageId: this.generateMessageId(),
          contents: [
            {
              kind: 'text',
              type: 'text',
              text: message,
            } as any,
          ],
        },
      ],
      metadata: options?.metadata,
    };

    // If tools are provided, handle tool execution
    if (options?.tools) {
      return await this.completeChatWithTools(request, options, signal);
    }

    const response = await this.createRunAndWait(request, signal);

    if (!response.output || response.output.length === 0) {
      return '';
    }

    // Extract text from first agent message
    const agentMessage = response.output.find((m) => m.role === 'agent');
    if (!agentMessage) {
      return '';
    }

    return this.extractText(agentMessage);
  }

  /**
   * Sends a structured message and returns the complete response
   * @param message The message to send (can include images, audio, etc.)
   * @param options Optional chat options
   * @param signal Optional abort signal for cancellation
   * @returns The agent's response message
   */
  async completeChatStructured(
    message: ChatMessage,
    options?: ChatOptions,
    signal?: AbortSignal
  ): Promise<ChatMessage> {
    const request: RunRequest = {
      agentId: options?.agentId,
      input: [message],
      metadata: options?.metadata,
    };

    const response = await this.createRunAndWait(request, signal);

    if (!response.output || response.output.length === 0) {
      return {
        role: 'agent' as ChatRole,
        messageId: this.generateMessageId(),
        contents: [],
      };
    }

    return (
      response.output.find((m) => m.role === 'agent') ?? {
        role: 'agent' as ChatRole,
        messageId: this.generateMessageId(),
        contents: [],
      }
    );
  }

  /**
   * Streams a message response with text chunks delivered via callback
   * @param message The message to send
   * @param onTextChunk Callback fired for each text chunk
   * @param options Optional chat options
   * @param signal Optional abort signal for cancellation
   */
  async streamChat(
    message: string,
    onTextChunk: (text: string) => void,
    options?: ChatOptions,
    signal?: AbortSignal
  ): Promise<void> {
    const request: RunRequest = {
      agentId: options?.agentId,
      input: [
        {
          role: 'user' as ChatRole,
          messageId: this.generateMessageId(),
          contents: [
            {
              kind: 'text',
              type: 'text',
              text: message,
            } as any,
          ],
        },
      ],
      metadata: options?.metadata,
    };

    let accumulatedText = '';

    for await (const event of this.streamRun(request, signal)) {
      // Handle different event types for text streaming
      if (event.eventType === 'message.delta' || event.eventType === 'message.updated') {
        const messageData = event.data as ChatMessage;
        if (messageData?.contents) {
          const textContent = messageData.contents.find(
            (c: any) => c.type === 'text'
          ) as any;

          if (textContent?.text) {
            // Calculate new text since last update
            const currentText = textContent.text;
            if (currentText.length > accumulatedText.length) {
              const newText = currentText.substring(accumulatedText.length);
              onTextChunk(newText);
              accumulatedText = currentText;
            }
          }
        }
      }
    }
  }

  /**
   * Creates a new conversation for maintaining state across multiple messages
   * @returns A conversation instance
   */
  createConversation(): IConversation {
    return new Conversation(
      this,
      undefined,
      this.config.enableLogging ?? false,
      this.config.logDirectory ?? 'logs/conversations'
    );
  }

  /**
   * Resumes an existing conversation using a thread ID
   * @param threadId The thread ID to resume
   * @returns A conversation instance
   */
  resumeConversation(threadId: string): IConversation {
    return new Conversation(
      this,
      threadId,
      this.config.enableLogging ?? false,
      this.config.logDirectory ?? 'logs/conversations'
    );
  }

  /**
   * Creates a run and waits for completion (internal method used by Conversation)
   * @internal
   */
  async createRunAndWait(request: RunRequest, signal?: AbortSignal): Promise<RunResponse> {
    const response = await this.lowLevelClient.runs.createAndWait(
      {
        agentId: request.agentId!,
        threadId: request.threadId,
        input: request.input,
        metadata: request.metadata,
      },
      {
        signal,
        timeout: 120000, // 2 minutes default
      }
    );

    return {
      runId: response.runId,
      threadId: response.threadId ?? '',
      status: response.status,
      output: response.output,
      error: response.error
        ? {
            code: response.error.code,
            message: response.error.message,
            details: response.error.details,
          }
        : undefined,
    };
  }

  /**
   * Streams a run and yields events (internal method used by Conversation)
   * @internal
   */
  async *streamRun(
    request: RunRequest,
    signal?: AbortSignal
  ): AsyncGenerator<StreamEvent, void, undefined> {
    // Create run via POST /runs
    const run = await this.lowLevelClient.runs.create(
      {
        agentId: request.agentId!,
        threadId: request.threadId,
        input: request.input,
        metadata: request.metadata,
      },
      { signal }
    );

    // Stream events via SSE
    const { SSEStream } = await import('./streaming');
    const baseUrl = (this.lowLevelClient as any).baseUrl;
    const authToken = (this.lowLevelClient as any).authToken;

    const stream = new SSEStream(`${baseUrl}/runs/${run.runId}/stream`, {
      authToken,
      autoReconnect: false,
    });

    // Convert SSE events to StreamEvent
    const eventQueue: StreamEvent[] = [];
    let streamComplete = false;
    let streamError: Error | null = null;

    stream.on('*', (event: any) => {
      const streamEvent = createStreamEvent(event.event, event);
      eventQueue.push(streamEvent);
    });

    stream.on('done', () => {
      streamComplete = true;
    });

    stream.on('error', (error: any) => {
      streamError = error;
      streamComplete = true;
    });

    // Handle abort signal
    const abortHandler = () => {
      stream.close();
      streamComplete = true;
    };
    signal?.addEventListener('abort', abortHandler);

    try {
      stream.connect();

      // Yield events as they arrive
      while (!streamComplete || eventQueue.length > 0) {
        if (signal?.aborted) {
          throw new Error('Stream aborted');
        }

        if (eventQueue.length > 0) {
          yield eventQueue.shift()!;
        } else if (!streamComplete) {
          // Wait a bit before checking again
          await new Promise((resolve) => setTimeout(resolve, 10));
        }
      }

      if (streamError) {
        throw streamError;
      }
    } finally {
      signal?.removeEventListener('abort', abortHandler);
      stream.close();
    }
  }

  /**
   * Handles tool execution automatically during streaming
   */
  private async completeChatWithTools(
    request: RunRequest,
    options: ChatOptions,
    signal?: AbortSignal
  ): Promise<string> {
    let resultText = '';

    for await (const event of this.streamRun(request, signal)) {
      if (event.eventType === 'message.delta' || event.eventType === 'message.updated') {
        const messageData = event.data as ChatMessage;
        if (messageData?.contents) {
          const textContent = messageData.contents.find(
            (c: any) => c.kind === 'text'
          ) as any;

          if (textContent?.text) {
            resultText = textContent.text;
          }
        }
      } else if (event.eventType === 'run.requires_action') {
        // TODO: Handle tool execution
        // Extract tool calls from the event
        // Execute tools using options.tools
        // Submit tool outputs back to the run
        const runData = event.data as any;
        if (runData && options.tools) {
          // This requires protocol-level support for submitting tool outputs
          // For now, tools need to be handled at a lower level
          console.warn('Tool execution during streaming not yet fully implemented');
        }
      }
    }

    return resultText;
  }

  /**
   * Extracts text content from a message
   */
  private extractText(message: ChatMessage): string {
    if (!message.contents || message.contents.length === 0) {
      return '';
    }

    const textContent = message.contents.find(
      (c: any) => c.kind === 'text'
    ) as any;

    return textContent?.text ?? '';
  }

  /**
   * Generates a unique message ID
   */
  private generateMessageId(): string {
    return `msg_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
  }

  /**
   * Gets the underlying low-level client for advanced operations
   * @returns The low-level AgentProtocolClient instance
   */
  get client(): AgentProtocolClient {
    return this.lowLevelClient;
  }
}
